tensorpipe/channel/cuda_basic/channel_impl.cc

```diff
@@ -30,165 +31,319 @@ ChannelImpl::ChannelImpl(
+    if (!op.done) {
       break;
     }
-
-    if (error_) {
-      op.descriptorCallback(error_, std::string());
-    } else {
-      CpuBuffer cpuBuffer{op.tmpBuffer.get(), op.length};
-      // Keep tmpBuffer alive until cpuChannel_ is done sending it over.
-      // TODO: This could be a lazy callback wrapper.
-      auto callback = callbackWrapper_(
-          [sequenceNumber{op.sequenceNumber},
-           tmpBuffer{std::move(op.tmpBuffer)}](ChannelImpl& impl) {
-            TP_VLOG(5) << "Channel " << impl.id_ << " is done sending buffer #"
-                       << sequenceNumber << " through CPU channel";
-          });
-      TP_VLOG(6) << "Channel " << id_ << " is sending buffer #"
-                 << op.sequenceNumber << " through CPU channel";
-      cpuChannel_->send(
-          cpuBuffer, std::move(op.descriptorCallback), std::move(callback));
+    if (op.callback) {
+      op.callback(error_);
+    }
+    if (!error_) {
+      sendChunkThroughCpuChannel(std::move(op));
     }
-
     sendOperations_.pop_front();
   }
+
+  if (error_) {
+    tryCleanup();
+  }
 }
 
 void ChannelImpl::recvImplFromLoop(
     uint64_t sequenceNumber,
     TDescriptor descriptor,
     CudaBuffer buffer,
     TRecvCallback callback) {
-  int deviceIdx = cudaDeviceForPointer(context_->getCudaLib(), buffer.ptr);
+  const size_t chunkLength = cudaHostAllocator_.getChunkLength();
+  const size_t numChunks = (buffer.length + chunkLength - 1) / chunkLength;
+  for (size_t offset = 0; offset < buffer.length; offset += chunkLength) {
+    recvOperations_.emplace_back();
+    Operation& op = recvOperations_.back();
+    op.sequenceNumber = sequenceNumber;
+    op.chunkId = offset / chunkLength;
+    op.numChunks = numChunks;
+    op.stream = buffer.stream;
+    op.cudaPtr = static_cast<uint8_t*>(buffer.ptr) + offset;
+    op.length = std::min(buffer.length - offset, chunkLength);
+    // Operations are processed in order, so we can afford to trigger the
+    // callback once the last operation is done.
+    if (op.chunkId == numChunks - 1) {
+      op.callback = std::move(callback);
+    }
+
+    TP_VLOG(6) << "Channel " << id_ << " is reading descriptor for chunk #"
+               << op.chunkId << " of " << op.numChunks << " for buffer #"
+               << op.sequenceNumber;
+    auto nopHolderIn = std::make_shared<NopHolder<std::string>>();
+    connection_->read(
+        *nopHolderIn,
+        callbackWrapper_([nopHolderIn, &op](ChannelImpl& impl) mutable {
+          TP_VLOG(6) << "Channel " << impl.id_
+                     << " is done reading descriptor for chunk #" << op.chunkId
+                     << " of " << op.numChunks << " for buffer #"
+                     << op.sequenceNumber;
+          std::string& descriptor = nopHolderIn->getObject();
+          impl.onRecvOpReadDescriptor(op, std::move(descriptor));
+        }));
+  }
+}
+
+void ChannelImpl::onRecvOpReadDescriptor(
+    Operation& op,
+    std::string descriptor) {
```

Comment:
The fact that here (and several more times below) we're always passing the descriptor along with the operation really gives a feeling that the descriptor should just be a field of the operation.